Crash in Http2StreamHandling

Introduced in #3873. I suspect that what I commented in https://github.com/akka/akka-http/pull/3873#discussion_r674645368 does not work completely (also not well enough tested, obviously...).

```
backend [ERROR] [07/26/2021 15:53:57.672] [default-akka.actor.default-dispatcher-9] [akka://default/system/Materializers/StreamSupervisor-0/flow-2-0-detacher] Error in stage [akka.http.impl.engine.http2.Http2Demux$$anon$1]: requirement failed
backend java.lang.IllegalArgumentException: requirement failed
backend         at scala.Predef$.require(Predef.scala:327)
backend         at akka.http.impl.engine.http2.Http2StreamHandling$OutStreamImpl.nextFrame(Http2StreamHandling.scala:689)
backend         at akka.http.impl.engine.http2.Http2StreamHandling$Sending.pullNextFrame(Http2StreamHandling.scala:334)
backend         at akka.http.impl.engine.http2.Http2StreamHandling$Sending.pullNextFrame$(Http2StreamHandling.scala:333)
backend         at akka.http.impl.engine.http2.Http2StreamHandling$HalfClosedRemoteSendingData.pullNextFrame(Http2StreamHandling.scala:498)
backend         at akka.http.impl.engine.http2.Http2StreamHandling.$anonfun$pullNextFrame$1(Http2StreamHandling.scala:140)
backend         at akka.http.impl.engine.http2.Http2StreamHandling.updateStateAndReturn(Http2StreamHandling.scala:155)
backend         at akka.http.impl.engine.http2.Http2StreamHandling.pullNextFrame(Http2StreamHandling.scala:140)
backend         at akka.http.impl.engine.http2.Http2StreamHandling.pullNextFrame$(Http2StreamHandling.scala:139)
backend         at akka.http.impl.engine.http2.Http2Demux$$anon$1.pullNextFrame(Http2Demux.scala:215)
backend         at akka.http.impl.engine.http2.Http2MultiplexerSupport$$anon$1$MultiplexerState.sendDataFrame(Http2Multiplexer.scala:160)
backend         at akka.http.impl.engine.http2.Http2MultiplexerSupport$$anon$1$MultiplexerState.sendDataFrame$(Http2Multiplexer.scala:158)
backend         at akka.http.impl.engine.http2.Http2MultiplexerSupport$$anon$1$WithSendableOutStreams.sendDataFrame(Http2Multiplexer.scala:252)
backend         at akka.http.impl.engine.http2.Http2MultiplexerSupport$$anon$1$WithSendableOutStreams.sendNext(Http2Multiplexer.scala:255)
backend         at akka.http.impl.engine.http2.Http2MultiplexerSupport$$anon$1$WaitingForNetworkToSendData$.onPull(Http2Multiplexer.scala:274)
backend         at akka.http.impl.engine.http2.Http2MultiplexerSupport$$anon$1.$anonfun$onPull$1(Http2Multiplexer.scala:123)
backend         at akka.http.impl.engine.http2.Http2MultiplexerSupport$$anon$1.updateState(Http2Multiplexer.scala:140)
backend         at akka.http.impl.engine.http2.Http2MultiplexerSupport$$anon$1.onPull(Http2Multiplexer.scala:123)
```

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala
@@ -330,23 +330,24 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper 
   trait Sending extends StreamState { _: Product =>
     protected def outStream: OutStream
 
-    override def pullNextFrame(maxSize: Int): (StreamState, PullFrameResult) = {
-      val frame = outStream.nextFrame(maxSize)
+    override def pullNextFrame(maxSize: Int): (StreamState, PullFrameResult) =
+      if (outStream.canSend) {
+        val frame = outStream.nextFrame(maxSize)
 
-      val res =
-        outStream.endStreamIfPossible() match {
-          case Some(trailer) =>
-            PullFrameResult.SendFrameAndTrailer(frame, trailer)
-          case None =>
-            PullFrameResult.SendFrame(frame, outStream.canSend)
-        }
+        val res =
+          outStream.endStreamIfPossible() match {
+            case Some(trailer) =>
+              PullFrameResult.SendFrameAndTrailer(frame, trailer)
+            case None =>
+              PullFrameResult.SendFrame(frame, outStream.canSend)
+          }
 
-      val nextState =
-        if (outStream.isDone) handleOutgoingEnded()
-        else this
+        val nextState =
+          if (outStream.isDone) handleOutgoingEnded()
+          else this
 
-      (nextState, res)
-    }
+        (nextState, res)
+      } else (this, PullFrameResult.NothingToSend)
 
     def handleWindowUpdate(windowUpdate: WindowUpdateFrame): StreamState = increaseWindow(windowUpdate.windowSizeIncrement)
 

--- a/akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala
+++ b/akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala
@@ -640,6 +640,25 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
         entityDataOut.sendComplete()
         network.expectDATA(TheStreamId, endStream = true, ByteString.empty)
       }
+      "keep sending entity data when WINDOW_UPDATE is received intermediately" inAssertAllStagesStopped new WaitingForResponseDataSetup {
+        val data1 = ByteString("abcd")
+        entityDataOut.sendNext(data1)
+
+        network.sendWINDOW_UPDATE(TheStreamId, 100)
+
+        val data2 = ByteString("efghij")
+        entityDataOut.sendNext(data2)
+
+        val (false, data) = network.expectDATAFrame(TheStreamId)
+        data shouldEqual data1 ++ data2
+
+        // now don't fail if there's demand on the line
+        network.plainDataProbe.request(1)
+        network.expectNoBytes(100.millis)
+
+        entityDataOut.sendComplete()
+        network.expectDATA(TheStreamId, endStream = true, ByteString.empty)
+      }
 
       "parse priority frames" inAssertAllStagesStopped new WaitingForResponseDataSetup {
         network.sendPRIORITY(TheStreamId, true, 0, 5)

